src/persist-client/src/operators/shard_source.rs

```diff
@@ -199,174 +239,144 @@ where
             .expect("could not open persist shard");
         let as_of = as_of.unwrap_or_else(|| read.since().clone());
 
-        // Eagerly yield the initial as_of. This makes sure that the output
-        // frontier of the `persist_source` closely tracks the `upper` frontier
-        // of the persist shard. It might be that the snapshot for `as_of` is
-        // not initially available yet, but this makes sure we already downgrade
-        // to it.
-        //
-        // Downstream consumers might rely on close frontier tracking for making
-        // progress. For example, the `persist_sink` needs to know the
-        // up-to-date upper of the output shard to make progress because it
-        // will only write out new data once it knows that earlier writes went
-        // through, including the initial downgrade of the shard upper to the
-        // `as_of`.
-        yield ListenEvent::Progress(as_of.clone());
-
         let subscription = read.subscribe(as_of.clone()).await;
-
         let mut subscription = subscription.unwrap_or_else(|e| {
             panic!(
                 "{}: {} cannot serve requested as_of {:?}: {:?}",
                 name_owned, shard_id, as_of, e
             )
         });
 
-        let mut done = false;
-        while !done {
-            while let Ok(leased_part) = consumed_part_rx.try_recv() {
-                subscription
-                    .return_leased_part(subscription.leased_part_from_exchangeable(leased_part));
-            }
-
-            for event in subscription.next().await {
-                if let ListenEvent::Progress(ref progress) = event {
-                    // If `until.less_equal(progress)`, it means that all subsequent batches will
-                    // contain only times greater or equal to `until`, which means they can be
-                    // dropped in their entirety. The current batch must be emitted, but we can
-                    // stop afterwards.
-                    if PartialOrder::less_equal(&until, progress) {
-                        done = true;
-                    }
+        let mut emitted_to_until = false;
+        let mut parts_buffer = vec![];
+        let mut batch_parts = vec![];
+
+        let mut lease_returner = subscription.clone_lease_returner();
```

Comment:
The subscription stream needs long-term _exclusive_ access (as in, this access lasts the entire lifetime of the select-loop) to the `subscription` object, which prevents any other access to it. This is required because the `Susbscription::next` futures are not cancel-safe, which means we must wrap them in a stream that holds a mutable reference to the underlying object forever, so we don't accidentally drop any non-cancel-safe futures in the `select!`

This means there has to be a separate object, with shared, interior-mutable access to the lease-return parts of the subscription object

Reply:
Interesting! I feel a bit sad about breaking our interface / imposing costs on all callers to solve what seems to be mostly a source problem, but I don't have any other ideas...

OOC, how did we determine `next` wasn't cancel safe?

Reply:
> Interesting! I feel a bit sad about breaking our interface / imposing costs on all callers to solve what seems to be mostly a source problem, but I don't have any other ideas...

It's definitely not the prettiest, though I think it was a bit of Dan's master plan to move `shard_source` into `persist-client` so that we _could_ muck with the internals of `Subscribe`, `ReadHandle` etc without bubbling those changes up to a fully public interface. https://materializeinc.slack.com/archives/C03P1EGB7A7/p1672954591719999

> OOC, how did we determine next wasn't cancel safe?

On a micro level, `Subscribe` mutates its own state between `await` points within `next()`. On a macro level, even if `Subscribe::next` were cancel safe, I don't think we'd want to be in the business of maintaining such a guarantee because it's so easy to make a mistake in the future, and soooo very hard to test.

Reply:
yeah, +1 on everything in Paul's last comment